Check for internal memory overlap in some indexing-type functions

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43423 Check for internal memory overlap in some indexing-type functions**
* #43422 TensorIteratorConfig: Check memory overlap by default

Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

peterbell10 committed:
Check for internal memory overlap in some indexing-type functions

[ghstack-poisoned]

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"




[ghstack-poisoned]

ngimel commented:
cc @neerajprad @fritzo for changes to distribution tests. My impression was that broadcasted tensors are routinely used in distributions, so requiring non-self-overlapping outputs could lead to significant memory penaly.

peterbell10 commented on torch/distributions/geometric.py:
```diff
@@ -88,7 +88,8 @@ def sample(self, sample_shape=torch.Size()):
-        value, probs = broadcast_all(value, self.probs.clone(memory_format=torch.contiguous_format))
+        value, probs = broadcast_all(value, self.probs)
+        probs = probs.clone(memory_format=torch.contiguous_format)
         probs[(probs == 1) & (value == 0)] = 0
```
>My impression was that broadcasted tensors are routinely used in distributions, so requiring non-self-overlapping outputs could lead to significant memory penaly.

This is a bug fix. If `probs` is broadcasted here, then this line can write to locations where `value != 0` if it happens to overlap with a location where `value` is zero. The full tensor size will be required in the following calculation anyway, so I think this is fine.

The same reasoning applies to the change in multinomial.

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

fritzo reviewed (approved):
LGTM thanks for fixing this!

@ngimel I believe the only place where broadcasting would help reduce memory footprint is in `MyDist(...),log_prob()`, but those already perform arithmetic operations like `(-probs).log1p()` so I believe max extra overhead of this PR would be +25%.

fritzo commented:
Can someone else please review the added `at::assert_no_overlap()`s?

zou3519 commented:
> Can someone else please review the added `at::assert_no_overlap()`s?

I can review it after I finish reviewing #43422

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

zou3519 reviewed (commented):
Some not-uncommon use cases of `masked_fill_` and `index_put_` are idempotent. For example, I've seen user code out in the wild that tries to zero-out nans that does the following, but I don't know how many users actually try to do this on broadcasted tensors.
```
x[torch.isnan(x)] = 0
y.masked_fill_(torch.isnan(y), 0)
```
The other operators in this PR look fine, though.

My opinion is to not change masked_fill_ and index_put_, at least not in this PR, because it is a common paradigm to do `x[x == blah] = value`. If we do decide we want to fix masked_fill_ and index_put_, I think we should consider that change to be BC-breaking and have a deprecation cycle.

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]

peterbell10 commented:
The original issue (#39639) was originally about `masked_fill_` specifically, so I don't want to ignore it completely here. So, I've lowered it to a deprecation warning. But I could move that to another PR as well if you prefer.

zou3519 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -291,6 +292,13 @@ Tensor & _index_put_impl_(Tensor & self, TensorList indices, const Tensor & valu
   }
+
+  if (at::has_internal_overlap(self) == MemOverlap::YES) {
+    TORCH_WARN("Use of index_put_ on expanded tensors is deprecated. "
+               "Please clone() the tensor before performing this operation.");
```
Another way to get here is via advanced indexing. We should talk about that as well in the error message so that the user can connect these two ops. "Use of index_put_ on expanded tensors is deprecated. Please clone() the tensor before performing this operation. You may also encounter this error message when assigning items to a tensor using advanced indexing, that is also deprecated"

peterbell10 replied:
Good idea. Added a note in the warning for `index_put_` and also `masked_fill_` as well.

zou3519 commented:
> The original issue (#39639) was originally about `masked_fill_` specifically, so I don't want to ignore it completely here. I've lowered it to a deprecation warning. But I could move that to another PR as well if you prefer.

Sure, let's do the deprecation warning in this PR then

zou3519 reviewed (commented):
Let's add a mention of advanced indexing into the index_put_ deprecation. After that this should be good to go

peterbell10 committed:
Update on "Check for internal memory overlap in some indexing-type functions"


Differential Revision: [D23298652](https://our.internmc.facebook.com/intern/diff/D23298652)

[ghstack-poisoned]
